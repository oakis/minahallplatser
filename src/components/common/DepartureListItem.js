import React, { PureComponent } from 'react';
import { View, TouchableWithoutFeedback } from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import moment from 'moment';
import { Text } from './';
import { colors } from '../style';

function formatTime(minutes) {
    const duration = moment.duration(minutes, 'minutes');
    return (minutes > 59) ? `${duration.hours()}h ${duration.minutes()}m` : minutes;
}

export class DepartureListItem extends PureComponent {
    render() {
        const { item, onPress } = this.props;
        const shouldShowMin = (item.timeFormat == 'minutes') ? true : false;
        const { clockLeft, clockNext } = item;
        const timeLeft = (item.timeLeft <= 0) ? 'Nu' : formatTime(item.timeLeft);
        const timeNext = (item.timeNext <= 0 && item.timeNext !== null) ? 'Nu' : formatTime(item.timeNext);
        const left = shouldShowMin ? timeLeft : clockLeft;
        const next = shouldShowMin ? timeNext : clockNext;
        const getFontColor = () => {
            if (!item.isLive) {
                return colors.warning;
            } else if (timeLeft == 'Nu') {
                return colors.danger;
            }
            return colors.default;
        };
        const height = 50;

        const styles = {
            listStyle: {
                flex: 1,
                backgroundColor: (item.index % 2) ? colors.alternative : colors.lightgrey,
                marginLeft: 0,
                paddingTop: 5,
                paddingBottom: 5,
                flexDirection: 'row',
                alignItems: 'center'
            },
            col1Style: {
                height: 40,
                width: 50,
                alignItems: 'center',
                justifyContent: 'center',
                paddingLeft: 7
            },
            col2Style: {
                flex: 1,
                paddingLeft: 10,
                justifyContent: 'center'
            },
            col3Style: {
                height,
                width: 50,
                alignItems: 'center',
                justifyContent: 'center',
                marginRight: 5
            },
            stopNumStyle: {
                flex: 1,
                height: 40,
                width: 40,
                backgroundColor: item.fgColor,
                borderWidth: 2,
                borderRadius: 3,
                alignItems: 'center',
                justifyContent: 'center',
            },
            stopNumText: {
                color: item.bgColor,
                textAlign: 'center',
                textAlignVertical: 'center',
                fontWeight: 'bold',
                fontSize: (item.sname.length > 3) ? 12 : 14,
            },
            departureStyle: {
                fontSize: (!shouldShowMin) ? 18 : (item.timeLeft > 59) ? 14 : 24,
                color: getFontColor()
            },
            nextDepStyle: {
                fontSize: 12
            },
            directionStyle: {
                fontWeight: 'bold'
            },
            viaStyle: {
                marginTop: -5,
                fontSize: 10
            },
            iconStyle: {
                marginLeft: 5,
                alignSelf: 'center'
            }
        };
        const { stopNumStyle, col1Style, col2Style, col3Style, stopNumText,
                departureStyle, nextDepStyle, directionStyle, listStyle, viaStyle, iconStyle } = styles;

        return (
            <TouchableWithoutFeedback
                onPress={onPress}
                delayLongPress={500}
            >
                <View style={listStyle}>
                    <View style={col1Style}>
                        <View style={stopNumStyle}>
                            <Text style={stopNumText}>{item.sname}</Text>
                        </View>
                    </View>

                    <View style={col2Style}>
                        <Text style={directionStyle}>{item.direction}</Text>
                        {(item.via) ? <Text style={viaStyle}>{item.via}</Text> : null}
                        <View style={{ flexDirection: 'row' }}>
                            <Text>LÃ¤ge {item.track}</Text>
                            {(Object.prototype.hasOwnProperty.call(item, 'accessibility')) ?
                                <Icon name={'wheelchair'} size={13} style={iconStyle}>
                                </Icon> :
                            null}
                        </View>
                    </View>

                    <View style={col3Style}>
                        <Text style={departureStyle}>{left}</Text>
                        <Text style={nextDepStyle}>{next}</Text>
                    </View>
                </View>
            </TouchableWithoutFeedback>
        );
    }
}
