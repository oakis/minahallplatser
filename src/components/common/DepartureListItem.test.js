import React from 'react';
import {render} from '@testing-library/react-native';
import {DepartureListItem} from './DepartureListItem';

const item = {
  sname: '16X',
  direction: 'Högsbo',
  clockLeft: '16:00',
  clockNext: '16:05',
  timeLeft: '3',
  timeNext: '8',
  isLive: true,
  fgColor: '#eee',
  bgColor: '#000',
  type: 'BUS',
  via: 'via Centralstationen',
  track: 'A',
  accessibility: 'wheelChair',
  journeyid: '123',
  global: true,
  local: true,
};

const item2 = {
  sname: 'SVART',
  direction: 'Högsbo',
  clockLeft: '16:00',
  clockNext: '16:05',
  timeLeft: '0',
  timeNext: '8',
  isLive: true,
  fgColor: '#eee',
  bgColor: '#000',
  type: 'BOAT',
  via: '',
  track: '',
  accessibility: null,
  journeyid: '123',
  index: 1,
  timeFormat: '',
  global: false,
  local: true,
};

const item3 = {
  sname: '1',
  direction: 'Centralstationen',
  clockLeft: '16:00',
  clockNext: '16:05',
  timeLeft: '60',
  timeNext: '8',
  isLive: false,
  fgColor: '#eee',
  bgColor: '#000',
  type: 'VAS',
  via: '',
  track: '',
  accessibility: null,
  journeyid: '123',
  timeFormat: 'minutes',
  global: true,
  local: false,
};

it('should match snapshot for extra lines', () => {
  const {toJSON} = render(
    <DepartureListItem
      item={item}
      onPress={jest.fn()}
      onLongPress={jest.fn()}
    />,
  );
  expect(toJSON()).toMatchSnapshot();
});

it('should match snapshot for boats', () => {
  const {toJSON} = render(
    <DepartureListItem
      item={item2}
      onPress={jest.fn()}
      onLongPress={jest.fn()}
    />,
  );
  expect(toJSON()).toMatchSnapshot();
});

it('should match snapshot for trains', () => {
  const {toJSON} = render(
    <DepartureListItem
      item={item3}
      onPress={jest.fn()}
      onLongPress={jest.fn()}
    />,
  );
  expect(toJSON()).toMatchSnapshot();
});
