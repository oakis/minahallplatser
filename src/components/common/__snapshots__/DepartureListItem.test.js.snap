// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot for boats 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "flexDirection": "row",
      "marginLeft": 0,
      "paddingBottom": 5,
      "paddingTop": 5,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 40,
        "justifyContent": "center",
        "paddingLeft": 7,
        "width": 50,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#eee",
          "borderRadius": 3,
          "borderWidth": 2,
          "flex": 1,
          "height": 40,
          "justifyContent": "center",
          "width": 40,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "System",
              "fontSize": 16,
            },
            Object {
              "color": "#000",
              "fontSize": 10,
              "fontWeight": "bold",
              "textAlign": "center",
              "textAlignVertical": "center",
            },
          ]
        }
      >
        SVART
      </Text>
      <MaterialIcon
        color="#000"
        name="directions-boat"
        size={15}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "paddingLeft": 10,
      }
    }
  >
    <View
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "System",
              "fontSize": 16,
            },
            Object {
              "fontWeight": "bold",
            },
          ]
        }
      >
        Högsbo
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "System",
              "fontSize": 16,
            },
            undefined,
          ]
        }
      >
        Läge 
        A
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 50,
        "justifyContent": "center",
        "marginRight": 5,
        "width": 56,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "System",
            "fontSize": 16,
          },
          Object {
            "color": "#FF6961",
            "fontSize": 18,
          },
        ]
      }
    >
      16:00
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "System",
            "fontSize": 16,
          },
          Object {
            "fontSize": 12,
          },
        ]
      }
    >
      16:05
    </Text>
  </View>
</View>
`;

exports[`should match snapshot for extra lines 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#efefef",
      "flex": 1,
      "flexDirection": "row",
      "marginLeft": 0,
      "paddingBottom": 5,
      "paddingTop": 5,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 40,
        "justifyContent": "center",
        "paddingLeft": 7,
        "width": 50,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#eee",
          "borderRadius": 3,
          "borderWidth": 2,
          "flex": 1,
          "height": 40,
          "justifyContent": "center",
          "width": 40,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "System",
              "fontSize": 16,
            },
            Object {
              "color": "#000",
              "fontSize": 14,
              "fontWeight": "bold",
              "textAlign": "center",
              "textAlignVertical": "center",
            },
          ]
        }
      >
        16X
      </Text>
      <MaterialIcon
        color="#000"
        name="directions-bus"
        size={15}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "paddingLeft": 10,
      }
    }
  >
    <View
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "System",
              "fontSize": 16,
            },
            Object {
              "fontWeight": "bold",
            },
          ]
        }
      >
        Högsbo
      </Text>
      <MaterialIcon
        name="public"
        size={13}
        style={
          Object {
            "alignSelf": "center",
            "color": "#779ECB",
            "marginLeft": 2,
            "marginTop": 2,
          }
        }
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "System",
            "fontSize": 16,
          },
          Object {
            "fontSize": 12,
            "marginTop": -5,
          },
        ]
      }
    >
      via Centralstationen
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "System",
              "fontSize": 16,
            },
            undefined,
          ]
        }
      >
        Läge 
        A
      </Text>
      <MaterialIcon
        name="accessible"
        size={13}
        style={
          Object {
            "alignSelf": "center",
            "marginLeft": 5,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 50,
        "justifyContent": "center",
        "marginRight": 5,
        "width": 56,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "System",
            "fontSize": 16,
          },
          Object {
            "color": "#000",
            "fontSize": 18,
          },
        ]
      }
    >
      16:00
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "System",
            "fontSize": 16,
          },
          Object {
            "fontSize": 12,
          },
        ]
      }
    >
      16:05
    </Text>
  </View>
</View>
`;

exports[`should match snapshot for trains 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#efefef",
      "flex": 1,
      "flexDirection": "row",
      "marginLeft": 0,
      "paddingBottom": 5,
      "paddingTop": 5,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 40,
        "justifyContent": "center",
        "paddingLeft": 7,
        "width": 50,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#eee",
          "borderRadius": 3,
          "borderWidth": 2,
          "flex": 1,
          "height": 40,
          "justifyContent": "center",
          "width": 40,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "System",
              "fontSize": 16,
            },
            Object {
              "color": "#000",
              "fontSize": 14,
              "fontWeight": "bold",
              "textAlign": "center",
              "textAlignVertical": "center",
            },
          ]
        }
      >
        1
      </Text>
      <MaterialIcon
        color="#000"
        name="train"
        size={15}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "paddingLeft": 10,
      }
    }
  >
    <View
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "System",
              "fontSize": 16,
            },
            Object {
              "fontWeight": "bold",
            },
          ]
        }
      >
        Centralstationen
      </Text>
      <MaterialIcon
        name="public"
        size={13}
        style={
          Object {
            "alignSelf": "center",
            "color": "#779ECB",
            "marginLeft": 2,
            "marginTop": 2,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "System",
              "fontSize": 16,
            },
            undefined,
          ]
        }
      >
        Läge 
        A
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 50,
        "justifyContent": "center",
        "marginRight": 5,
        "width": 56,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "System",
            "fontSize": 16,
          },
          Object {
            "color": "#FFB347",
            "fontSize": 14,
          },
        ]
      }
    >
      1h 0m
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "System",
            "fontSize": 16,
          },
          Object {
            "fontSize": 12,
          },
        ]
      }
    >
      8
    </Text>
  </View>
</View>
`;
